util/dvsim/Testplan.py

```diff
@@ -0,0 +1,686 @@
+                "    text-align: left;\n"
+                "    vertical-align: middle;\n"
+                "    display: table;\n"
+                "    font-size: smaller;\n"
+                "}\n"
+                "table.dv th, td {\n"
+                "    border: 1px solid black;\n"
+                "}\n"
+                "</style>\n")
+
+    def __str__(self):
+        lines = [f"Name: {self.name}"]
+        lines += [
+            "\nTestpoints:\n" + "\n".join([f"{t}" for t in self.testpoints])
+        ]
+        lines += [
+            "\nCovergroups:\n" + "\n".join([f"{c}" for c in self.covergroups])
+        ]
+        return "\n".join(lines)
+
+    def __init__(self, filename, repo_top=None, name=None):
+        """Initialize the testplan.
+
+        filename is the HJson file that captures the testplan.
+        repo_top is an optional argument indicating the path to top level repo
+        / project directory. It is used with filename arg.
+        name is an optional argument indicating the name of the testplan / DUT.
+        It overrides the name set in the testplan HJson.
+        """
+        self.name = None
+        self.testpoints = []
+        self.covergroups = []
+        self.test_results_mapped = False
+
+        if filename:
+            self._parse_testplan(filename, repo_top)
+
+        if name is not None:
+            self.name = name
+
+        if self.name is None:
+            print("Error: the testplan 'name' is not set!")
+            sys.exit(1)
+
+        # Represents current progress towards each milestone. Milestone = N.A.
+        # is used to indicate the unmapped tests.
+        self.progress = {}
+        for key in Testpoint.milestones:
+            self.progress[key] = {
+                "written": 0,
+                "total": 0,
+                "progress": 0.0,
+            }
+
+    def _parse_testplan(self, filename, repo_top=None):
+        '''Parse testplan Hjson file and create the testplan elements.
+
+        It creates the list of testpoints and covergroups extracted from the
+        file.
+
+        filename is the path to the testplan file written in HJson format.
+        repo_top is an optional argument indicating the path to repo top.
+
+        '''
+        if repo_top is None:
+            # Assume dvsim's original location: $REPO_TOP/util/dvsim.
+            self_path = os.path.dirname(os.path.realpath(__file__))
+            repo_top = os.path.abspath(
+                os.path.join(self_path, os.pardir, os.pardir))
+
+        obj = Testplan._parse_hjson(filename)
+        imported_testplans = obj.get("import_testplans", [])
+        for imported_testplan in imported_testplans:
+            path = os.path.join(repo_top, imported_testplan)
+            obj = _merge_dicts(obj, self._parse_hjson(path))
+
+        self.name = obj.get("name")
+
+        # TODO: remove workaround once all testplans are updated.
+        testpoints = obj.get("testpoints", [])
```

Comment:
Should this emit some warning if either "testpoints" or "covergroups" are missing?